4-projects/modules/ml_single_project/main.tf

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  enabled_apis = distinct(concat(var.activate_apis, ["billingbudgets.googleapis.com"]))
+  env_code     = element(split("", var.environment), 0)
+  source_repos = setintersection(
+    toset(keys(var.app_infra_pipeline_service_accounts)),
+    toset(keys(var.sa_roles))
+  )
+  pipeline_roles = var.enable_cloudbuild_deploy ? flatten([
+    for repo in local.source_repos : [
+      for role in var.sa_roles[repo] :
+      {
+        repo = repo
+        role = role
+        sa   = var.app_infra_pipeline_service_accounts[repo]
+      }
+    ]
+  ]) : []
+
+  pipeline_kms_sas = var.enable_cloudbuild_deploy ? flatten([
+    for repo in keys(var.sa_roles) : [
+      var.app_infra_pipeline_service_accounts[repo]
+    ]
+  ]) : []
+
+  network_user_role = var.enable_cloudbuild_deploy ? flatten([
+    for repo in local.source_repos : [
+      for subnet in var.shared_vpc_subnets :
+      {
+        repo   = repo
+        subnet = element(split("/", subnet), index(split("/", subnet), "subnetworks", ) + 1, )
+        region = element(split("/", subnet), index(split("/", subnet), "regions") + 1, )
+        sa     = var.app_infra_pipeline_service_accounts[repo]
+      }
+    ]
+  ]) : []
+}
+
+module "project" {
+  source  = "terraform-google-modules/project-factory/google"
+  version = "~> 14.1"
+
+  random_project_id        = true
+  random_project_id_length = 4
+  activate_apis            = local.enabled_apis
+  name                     = var.project_name
+  org_id                   = var.org_id
+  billing_account          = var.billing_account
+  folder_id                = var.folder_id
+  default_service_account  = var.default_service_account
+
+  svpc_host_project_id = var.shared_vpc_host_project_id
+  shared_vpc_subnets   = var.shared_vpc_subnets # Optional: To enable subnetting, replace to "module.networking_project.subnetwork_self_link"
+
+  vpc_service_control_attach_enabled = var.vpc_service_control_attach_enabled
```

Comment:
Not sure if it is applicable here, but it could be nice to have `vpc_service_control_attach_dry_run` in case use want to use VPC SC in a dry run mode instead.